lib/libzfs/libzfs_util.c

```diff
@@ -661,26 +686,80 @@ libzfs_run_process(const char *path, char *argv[], int flags)
 	return (-1);
 }
 
-int
+/*
+ * Verify the required ZFS_DEV device is available and optionally attempt
+ * to load the ZFS modules.  Under normal circumstances the modules
+ * should already have been loaded by some external mechanism.
+ *
+ * Environment variables:
+ * - ZFS_MODULE_LOADING="YES|yes|ON|on" - Attempt to load modules.
+ * - ZFS_MODULE_TIMEOUT="<seconds>"     - Seconds to wait for ZFS_DEV
+ */
+static int
 libzfs_load_module(const char *module)
 {
 	char *argv[4] = {"/sbin/modprobe", "-q", (char *)module, (char *)0};
+	char *load_str, *timeout_str;
+	long timeout = 10; /* seconds */
+	long busy_timeout = 10; /* milliseconds */
+	int load = 0, fd;
+	hrtime_t start;
+
+	/* Optionally request module loading */
+	if (!libzfs_module_loaded(module)) {
+		load_str = getenv("ZFS_MODULE_LOADING");
+		if (load_str &&
+		    (!strncasecmp(load_str, "YES", strlen("YES")) ||
+		    (!strncasecmp(load_str, "ON", strlen("ON")))))
+			load = 1;
+
+		if (load && libzfs_run_process("/sbin/modprobe", argv, 0))
+			return (ENOEXEC);
+	}
 
-	if (libzfs_module_loaded(module))
-		return (0);
+	/* Module loading is synchronous it must be available */
+	if (!libzfs_module_loaded(module))
+		return (ENXIO);
+
+	/*
+	 * Device creation by udev is asynchronous and waiting may be
+	 * required.  Busy wait for 10ms and then fall back to polling every
+	 * 10ms for the allowed timeout (default 10s, max 10m).  This is
+	 * done to optimize for the common case where the device is
+	 * immediately available and to avoid penalizing the possible
+	 * case where udev is slow or unable to create the device.
+	 */
+	timeout_str = getenv("ZFS_MODULE_TIMEOUT");
+	if (timeout_str)
+		timeout = MIN(strtol(timeout_str, NULL, 0), (10 * 60));
+
+	start = gethrtime();
+	while (NSEC2MSEC(gethrtime() - start) < (timeout * MILLISEC)) {
+		fd = open(ZFS_DEV, O_RDWR);
+		if (fd >= 0) {
+			(void) close(fd);
+			return (0);
+		} else if (errno == ENOENT) {
+			if (NSEC2MSEC(gethrtime() - start) > busy_timeout)
+				usleep(10 * MILLISEC);
```

Comment:
I actually intentionally dropped it because we're only going to be spinning briefly.  But I can add it back, it does appear to be POSIX so there aren't compatibility concerns.